File: src/main/java/me/zeroeightsix/kami/module/modules/chat/DiscordNotifs.java
package me.zeroeightsix.kami.module.modules.chat;

import com.mrpowergamerbr.temmiewebhook.DiscordMessage;
import com.mrpowergamerbr.temmiewebhook.TemmieWebhook;
import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.KamiMod;
import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.event.events.GuiScreenEvent;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.event.events.ServerConnectedEvent;
import me.zeroeightsix.kami.event.events.ServerDisconnectedEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.module.modules.client.InfoOverlay;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.TimeUtil;
import net.minecraft.client.gui.GuiDisconnected;
import net.minecraft.client.multiplayer.GuiConnecting;
import net.minecraft.client.multiplayer.ServerData;
import net.minecraft.network.play.server.SPacketChat;

import static me.zeroeightsix.kami.KamiMod.EVENT_BUS;
import static me.zeroeightsix.kami.KamiMod.MODULE_MANAGER;
import static me.zeroeightsix.kami.util.MessageDetectionHelper.*;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendErrorMessage;

/**
 * @author dominikaaaa
 * Created by dominikaaaa on 26/03/20
 * Updated by dominikaaaa on 28/03/20
 */
@Module.Info(
        name = "DiscordNotifs",
        category = Module.Category.CHAT,
        description = "Sends your chat to a set Discord channel",
        showOnArray = Module.ShowOnArray.OFF,
        alwaysListening = true
)
public class DiscordNotifs extends Module {
    private Setting<Boolean> timeout = register(Settings.b("Timeout", true));
    private Setting<Integer> timeoutTime = register(Settings.integerBuilder().withName("Seconds").withMinimum(0).withMaximum(120).withValue(10).withVisibility(v -> timeout.getValue()).build());
    private Setting<Boolean> time = register(Settings.b("Timestamp", true));
    private Setting<Boolean> importantPings = register(Settings.b("Important Pings", false));
    private Setting<Boolean> disconnect = register(Settings.b("Disconnect Msgs", true));
    private Setting<Boolean> all = register(Settings.b("All Messages", false));
    private Setting<Boolean> queue = register(Settings.booleanBuilder("Queue Position").withValue(true).withVisibility(v -> !all.getValue()).build());
    private Setting<Boolean> restart = register(Settings.booleanBuilder("Restart Msgs").withValue(true).withVisibility(v -> !all.getValue()).build());
    private Setting<Boolean> direct = register(Settings.booleanBuilder("Received DMs").withValue(true).withVisibility(v -> !all.getValue()).build());
    private Setting<Boolean> directSent = register(Settings.booleanBuilder("Send DMs").withValue(true).withVisibility(v -> !all.getValue()).build());
    public Setting<String> url = register(Settings.s("URL", "unchanged"));
    public Setting<String> pingID = register(Settings.s("Ping ID", "unchanged"));
    public Setting<String> avatar = register(Settings.s("Avatar", KamiMod.GITHUB_LINK + "assets/raw/assets/assets/icons/kami.png"));

    private static ServerData cServer;

    /* Listeners to send the messages */
    @EventHandler
    public Listener<PacketEvent.Receive> listener0 = new Listener<>(event -> {
        if (mc.player == null || isDisabled()) return;
        if (!(event.getPacket() instanceof SPacketChat)) return;

        SPacketChat sPacketChat = (SPacketChat) event.getPacket();
        String message = sPacketChat.getChatComponent().getUnformattedText();
        if (timeout(message) && shouldSend(all.getValue(), restart.getValue(), direct.getValue(), directSent.getValue(), queue.getValue(), importantPings.getValue(), message)) {
            sendMessage(getPingID(message) + getMessageType(direct.getValue(), directSent.getValue(), message, getServer()) + getTime() + message, avatar.getValue());
        }
    });

    @EventHandler
    public Listener<ServerConnectedEvent> listener1 = new Listener<>(event -> {
        if (isDisabled()) return;
        if (!disconnect.getValue()) return;
        sendMessage(getPingID("KamiBlueMessageType1") + getTime() + getMessageType(direct.getValue(), directSent.getValue(), "KamiBlueMessageType1", getServer()), avatar.getValue());
    });

    @EventHandler
    public Listener<ServerDisconnectedEvent> listener2 = new Listener<>(event -> {
        if (isDisabled()) return;
        if (!disconnect.getValue()) return;
        sendMessage(getPingID("KamiBlueMessageType2") + getTime() + getMessageType(direct.getValue(), directSent.getValue(), "KamiBlueMessageType2", getServer()), avatar.getValue());
    });

    /* Getters for messages */
    private static long startTime = 0;
    private boolean timeout(String message) {
        if (!timeout.getValue()) return true;
        else if (isRestart(restart.getValue(), message) || isDirect(direct.getValue(), message) || isDirectOther(directSent.getValue(), message)) return true;
        if (startTime == 0) startTime = System.currentTimeMillis();
        if (startTime + (timeoutTime.getValue() * 1000) <= System.currentTimeMillis()) { // 1 timeout = 1 second = 1000 ms
            startTime = System.currentTimeMillis();
            return true;
        }
        return false;
    }

    /* Text formatting and misc methods */
    private String getPingID(String message) {
        if (isRestart(restart.getValue(), message) || isDirect(direct.getValue(), message) || isDirectOther(directSent.getValue(), message) || isImportantQueue(importantPings.getValue(), message)) return formatPingID();
        else if ((message.equals("KamiBlueMessageType1")) || (message.equals("KamiBlueMessageType2"))) return formatPingID();
        else return "";
    }

    private String formatPingID() {
        if (!importantPings.getValue()) return "";
        else return "<@!" + pingID.getValue() + ">: ";
    }

    private String getServer() {
        if (cServer == null) return "the server";
        else return cServer.serverIP;
    }

    private String getTime() {
        if (!time.getValue()) return "";
        InfoOverlay info = MODULE_MANAGER.getModuleT(InfoOverlay.class);
        return "[" + TimeUtil.getFinalTime(info.timeUnitSetting.getValue(), info.timeTypeSetting.getValue(), info.doLocale.getValue()) + "] ";
    }

    private void sendMessage(String content, String avatarUrl) {
        TemmieWebhook tm = new TemmieWebhook(url.getValue());
        DiscordMessage dm = new DiscordMessage(KamiMod.MODNAME + " " + KamiMod.MODVER, content, avatarUrl);
        tm.sendMessage(dm);
    }

    /* Always on status code */
    public void onUpdate() {
        if (isDisabled()) return;
        if (url.getValue().equals("unchanged")) {
            sendErrorMessage(getChatName() + "You must first set a webhook url with the '&7" + Command.getCommandPrefix() + "discordnotifs&r' command");
            disable();
        }
        else if (pingID.getValue().equals("unchanged") && importantPings.getValue()) {
            sendErrorMessage(getChatName() + "For Pings to work, you must set a Discord ID with the '&7" + Command.getCommandPrefix() + "discordnotifs&r' command");
            disable();
        }
    }

    @EventHandler
    public Listener<GuiScreenEvent.Closed> serverConnectedEvent = new Listener<>(event -> {
        if (isEnabled() && event.getScreen() instanceof GuiConnecting) {
            cServer = mc.currentServerData;
            EVENT_BUS.post(new ServerConnectedEvent());
        }
    });

    @EventHandler
    public Listener<GuiScreenEvent.Displayed> serverDisconnectedEvent = new Listener<>(event -> {
        if (isEnabled() && event.getScreen() instanceof GuiDisconnected && (cServer != null || mc.currentServerData != null)) {
            EVENT_BUS.post(new ServerDisconnectedEvent());
        }
    });
}

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/DiscordNotifs.kt
package me.zeroeightsix.kami.module.modules.chat

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import me.zeroeightsix.kami.KamiMod
import me.zeroeightsix.kami.command.CommandManager
import me.zeroeightsix.kami.event.events.ConnectionEvent
import me.zeroeightsix.kami.event.events.PacketEvent
import me.zeroeightsix.kami.module.Category
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.util.TickTimer
import me.zeroeightsix.kami.util.TimeUnit
import me.zeroeightsix.kami.util.text.*
import me.zeroeightsix.kami.util.threads.defaultScope
import me.zeroeightsix.kami.util.threads.safeListener
import net.minecraft.network.play.server.SPacketChat
import net.minecraftforge.fml.common.gameevent.TickEvent
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.kamiblue.event.listener.listener

internal object DiscordNotifs : Module(
    name = "DiscordNotifs",
    category = Category.CHAT,
    description = "Sends your chat to a set Discord channel"
) {
    private val timeout by setting("Timeout", true)
    private val timeoutTime by setting("Seconds", 10, 0..120, 5, { timeout })
    private val time by setting("Timestamp", true)
    private val importantPings by setting("ImportantPings", false)
    private val connectionChange by setting("ConnectionChange", true, description = "When you get disconnected or reconnected to the server")
    private val all by setting("AllMessages", false)
    private val direct by setting("DMs", true, { !all })
    private val queue by setting("QueuePosition", true, { !all })
    private val restart by setting("Restart", true, { !all }, description = "Server restart notifications")

    val url = setting("URL", "unchanged")
    val pingID = setting("PingID", "unchanged")
    val avatar = setting("Avatar", "${KamiMod.GITHUB_LINK}/assets/raw/assets/assets/icons/kamiGithub.png")

    private const val username = "${KamiMod.NAME} ${KamiMod.VERSION}"
    private val server: String get() = mc.currentServerData?.serverIP ?: "the server"
    private val timer = TickTimer(TimeUnit.SECONDS)

    /* Listeners to send the messages */
    init {
        listener<PacketEvent.Receive> {
            if (mc.player == null || it.packet !is SPacketChat) return@listener
            val message = it.packet.chatComponent.unformattedText
            if (timeout(message) && shouldSend(message)) {
                sendMessage(getPingID(message) + getMessageType(message, server) + getTime() + message)
            }
        }

        listener<ConnectionEvent.Connect> {
            if (!connectionChange) return@listener
            sendMessage(getPingID("KamiBlueMessageType1") + getTime() + getMessageType("KamiBlueMessageType1", server))
        }

        listener<ConnectionEvent.Disconnect> {
            if (!connectionChange) return@listener
            sendMessage(getPingID("KamiBlueMessageType2") + getTime() + getMessageType("KamiBlueMessageType2", server))
        }

        /* Always on status code */
        safeListener<TickEvent.ClientTickEvent> {
            if (url.value == "unchanged") {
                MessageSendHelper.sendErrorMessage(chatName + " You must first set a webhook url with the " +
                    formatValue("${CommandManager.prefix}discordnotifs") +
                    " command")
                disable()
            } else if (pingID.value == "unchanged" && importantPings) {
                MessageSendHelper.sendErrorMessage(chatName + " For Pings to work, you must set a Discord ID with the " +
                    formatValue("${CommandManager.prefix}discordnotifs") +
                    " command")
                disable()
            }
        }
    }

    private fun shouldSend(message: String): Boolean {
        return all
            || direct && MessageDetection.Direct.ANY detect message
            || restart && MessageDetection.Server.RESTART detect message
            || queue && MessageDetection.Server.QUEUE detect message
    }

    private fun getMessageType(message: String, server: String): String {
        if (direct && MessageDetection.Direct.RECEIVE detect message) return "You got a direct message!\n"
        if (direct && MessageDetection.Direct.SENT detect message) return "You sent a direct message!\n"
        if (message == "KamiBlueMessageType1") return "Connected to $server"
        return if (message == "KamiBlueMessageType2") "Disconnected from $server" else ""
    }

    private fun timeout(message: String) = !timeout
        || restart && MessageDetection.Server.RESTART detect message
        || direct && MessageDetection.Direct.ANY detect message
        || timer.tick(timeoutTime.toLong())

    /* Text formatting and misc methods */
    private fun getPingID(message: String) = if (message == "KamiBlueMessageType1"
        || message == "KamiBlueMessageType2"
        || direct && MessageDetection.Direct.ANY detect message
        || restart && MessageDetection.Server.RESTART detect message
        || importantPings && MessageDetection.Server.QUEUE_IMPORTANT detect message) formatPingID()
    else ""

    private fun formatPingID(): String {
        return if (!importantPings) "" else "<@!${pingID.value}>: "
    }

    private fun getTime() =
        if (!time) ""
        else ChatTimestamp.time

    private fun sendMessage(content: String) {
        val jsonType = "application/json; charset=utf-8".toMediaType()

        // todo json dsl
        val body = "{\"username\": \"$username\", \"content\": \"$content\", \"avatar_url\": \"${avatar.value}\"}".toRequestBody(jsonType)
        val request: Request = Request.Builder()
            .url(url.value)
            .addHeader("Content-Type", "application/json")
            .post(body)
            .build()

        defaultScope.launch(Dispatchers.IO) {
            // todo probably make some DSL for this lol
            KamiMod.getHttpClient().newCall(request).execute().use { response ->
                val responseBody = response.body?.string().toString()
                if (responseBody.isBlank()) return@launch // Returns 204 empty normally. We want to warn log any non-empty responses.
                KamiMod.LOG.warn("DiscordNotifs OkHttp Request\n${responseBody}")
            }
        }
    }
}